### PR TITLE
BindPython ABI

### PR DESCRIPTION
**Pull request by Vipul-Cariappa**

### CODE REVIEW

**Comment by certik** [comment]
The right solution here is to have an ASR->ASR pass that takes the `bind(Python)` and replaces it with calls to Python C/API. This should not be hardwired into all the backends.

If you think you can get something working with relatively short code, you can finish the PR, but I think having a pass is a better solution long term.

**Comment by Vipul-Cariappa** [comment]
> The right solution here is to have an ASR->ASR pass that takes the bind(Python) and replaces it with calls to Python C/API. This should not be hardwired into all the backends.

Yes. I am implementing this using an ASR to ASR pass. The `python_bind.py` implements the ASR to ASR pass.

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:4479]: Why are these needed in AST->ASR?

Ideally the initial ASR has nothing to do with Python C/API. Then the ASR->ASR pass translates `bind(python)` into C/API calls in ASR.

**Comment by certik** [comment]
Otherwise it looks like the right approach, it seems you are doing the type translation to/from Python in the ASR->ASR pass already, which is perfect.

File: src/libasr/pass/python_bind.cpp
Comment (Shaikh-Ubaid) [inline src/libasr/pass/python_bind.cpp:360]: Any reason to make the `abi` as `BindC`? I think we should be making it `Source`?

File: src/libasr/pass/python_bind.cpp
Comment (Shaikh-Ubaid) [inline src/libasr/pass/python_bind.cpp:27]: Why do we need to do `Cast_t` here (and the rest of the types below)? Isn't the returned value by the specific function (in this case `PyLong_AsLongLong`) already of the desired type?

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (Shaikh-Ubaid) [reply src/lpython/semantics/python_ast_to_asr.cpp:4479]: Exactly, even I was confused in my last meet with Vipul. Looking into the changes, I think I understand what he is doing here.

He has a new runtime module `cpython_bindings.py` which contains the declarations to the `C` functions in `Python.h` that we need. In the above change, he is just creating `AST` code for the following:

```
from cpython_bindings import Py_Initialize, Py_IsInitialized, PyRun_SimpleString, Py_DecodeLocale, PySys_SetArgv, Py_FinalizeEx, PyUnicode_FromString, PyUnicode_AsUTF8AndSize, PyImport_Import, Py_DecRef, Py_IncRef, PyObject_GetAttrString, PyTuple_New, PyTuple_SetItem, PyObject_CallObject, PyLong_AsLongLong, PyLong_AsUnsignedLongLong, PyLong_FromLongLong, PyLong_FromUnsignedLongLong, PyFloat_FromDouble, PyFloat_AsDouble, PyBool_FromLong, PyObject_IsTrue
```

When he visits the `AST` code, `lpython` reads/loads the `runtime/cpython_bindings.py` and automatically generates the declarations for the above `BindC` functions.

@Vipul-Cariappa  Initially this approach sounded interesting to me, but now after going through the changes, it seems like we should instead write the ASR code for these directly in the `python_bind` pass. I think if we simply extend the current `pybind_funcs` array with the parameter types and return type info (for example specifying types like here https://github.com/lcompilers/lpython/blob/9374feb17362d8046db1e4060d633f7ff9887100/src/libasr/codegen/asr_to_wasm.cpp#L206), we should be able to run a loop and generate the ASR nodes for declaring these. Then, we ideally would not be making any changes in the AST->ASR level. And we can also remove the `runtime/cpython_bindings.py` file.

**Review by Shaikh-Ubaid — COMMENTED** [review]
Great work, Vipul! Shared some comments. Rest looks good.

File: src/libasr/pass/python_bind.cpp
Comment (Vipul-Cariappa) [reply src/libasr/pass/python_bind.cpp:177]: Yes, the `bindpy_05.py` example does not require this. But it is required for other tests example `bindpy_01.py`. 

We append the current path to CPython's module search path. If we do not do this, CPython will fail with `ModuleNotFoundError`, because it only looks for modules installed using pip or the builtins/CPython standard library.

File: src/libasr/pass/python_bind.cpp
Comment (Vipul-Cariappa) [reply src/libasr/pass/python_bind.cpp:27]: The signature of `PyLong_AsLongLong` is
`long long PyLong_AsLongLong(PyObject*)`. i.e. 
`i64 PyLong_AsLongLong(CPtr)`.

Python does not have `PyLongAsInt` that returns an `int` (`i32`).

The `Cast_t` is required in the following situation.
Assume the following PythonBind function
```
@pythoncall(module="mypylib")
def f(a: i32) -> i32:
    pass
```
or
```
@pythoncall(module="mypylib")
def f(a: i8) -> i8:
    pass
```
The `Cast_t` casts the result of `PyLong_AsLongLong` from `i64` to `i32` or `i8` according to the actual type required.

File: src/libasr/pass/python_bind.cpp
Comment (Vipul-Cariappa) [reply src/libasr/pass/python_bind.cpp:360]: Ok. Will change it to Source in the next commit.

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (Vipul-Cariappa) [reply src/lpython/semantics/python_ast_to_asr.cpp:4479]: Ok. I will update the code to generate the required function declarations in the `python_bind` pass.

File: src/libasr/pass/python_bind.cpp
Comment (Shaikh-Ubaid) [reply src/libasr/pass/python_bind.cpp:27]: Makes sense. Thanks.

File: src/libasr/pass/python_bind.cpp
Comment (Shaikh-Ubaid) [reply src/libasr/pass/python_bind.cpp:177]: I think https://github.com/lcompilers/lpython/blob/9374feb17362d8046db1e4060d633f7ff9887100/integration_tests/bindpy_05.py#L72-L73 should help. I believe it adds the current directory `.` to the path.

Comment (Vipul-Cariappa) [reply src/libasr/pass/python_bind.cpp]: I can try this.

What does `pointer(argv1, i64)` do? What does the `pointer` function do?

Comment (Vipul-Cariappa) [reply src/libasr/pass/python_bind.cpp]: I figured it out.

I am no longer using `PyRun_SimpleString` and appending the path. I am using `Py_DecodeLocale` and `PySys_SetArgv`.

File: src/libasr/pass/python_bind.cpp
Comment (Vipul-Cariappa) [inline src/libasr/pass/python_bind.cpp:489]: I was thinking about moving these functions to `asr_utils.cpp`.
These are generic functions that can be used to create new `BindC` `Interface` function declarations.

Any thoughts? @certik @Shaikh-Ubaid.

Comment (Shaikh-Ubaid) [reply src/libasr/pass/python_bind.cpp]: Sounds good to me. Also see/consider if `pass_utils` can be a better option.

Comment (Vipul-Cariappa) [reply src/libasr/pass/python_bind.cpp]: I feel `asr_utils.cpp` is the better option. In the future, if we need to generate function declarations for using any C function, we could just use these functions. Putting it in `asr_utils.cpp` will let us use it in the AST to ASR step also (if necessary).

**Comment by Vipul-Cariappa** [comment]
@Shaikh-Ubaid, I believe if the CI passes and these changes look good to you, then we can merge.

File: src/libasr/pass/python_bind.cpp
Comment (Shaikh-Ubaid) [inline src/libasr/pass/python_bind.cpp:415]: I think `included_cpython_funcs` is not needed. We can simply call ` ASRUtils::declare_functions(al, fns, *l, unit.m_symtab, "Python.h")` here. We skip the whole `python_bind` pass if `--enable-cpython` is not provided.


Also suggested here https://github.com/lcompilers/lpython/pull/2796#discussion_r1704413641.

File: src/libasr/pass/python_bind.cpp
Comment (Shaikh-Ubaid) [inline src/libasr/pass/python_bind.cpp:459]: Minor: 
```suggestion

}  // namespace LCompilers
```

**Review by Shaikh-Ubaid — APPROVED** [review]
Shared one comment here https://github.com/lcompilers/lpython/pull/2796/files#r1713010953. I think its good to merge after that. Thanks for this @Vipul-Cariappa!

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp]: > Initially this approach sounded interesting to me, but now after going through the changes, it seems like we should instead write the ASR code for these directly in the `python_bind` pass.

Yes, exactly.

**Review by certik — APPROVED** [review]
Great job. After @Shaikh-Ubaid is happy, this can be merged. It looks clean and small, the pass has less than 460 lines and it does everything. I think this is the way to go.

Is this now used by default for all Python bindings, meaning all our existing tests pass with this new approach?

File: src/libasr/pass/python_bind.cpp
Comment (Vipul-Cariappa) [reply src/libasr/pass/python_bind.cpp:415]: The `--enable-cpython` flag sets `CompilerOptions.enable_cpython` to true. We do not have access to `CompilerOptions` within an ASR pass. We only have access to `PassOptions`.
Should I move `enable_cpython` to `PassOptions` then? Or have it in both places in `CompilerOptions` and `PassOptions`, so that we don't break any other code.

**Comment by Vipul-Cariappa** [comment]
@certik
The tests that use arrays (i.e. call into a CPython function that accepts arrays or returns an array) are skipped.
Conversion from LPython arrays to CPython arrays (numpy array) needs to be implemented. I will do it in the next PR.

File: src/libasr/pass/python_bind.cpp
Comment (Shaikh-Ubaid) [reply src/libasr/pass/python_bind.cpp:415]: We can move it to pass options.

Comment (Vipul-Cariappa) [reply src/libasr/pass/python_bind.cpp]: I have made the changes. Hope I understood it correctly and this is what you wanted.

**Review by Shaikh-Ubaid — APPROVED** [review]
Perfect, thanks Vipul! It looks great! I appreciate it.

**Pull request merged**
